iron-image.html

```diff
@@ -224,70 +250,76 @@
           type: Number,
           value: null
         },
+      },
 
-        _placeholderBackgroundUrl: {
-          type: String,
-          computed: '_computePlaceholderBackgroundUrl(preload,placeholder)',
-          observer: '_placeholderBackgroundUrlChanged'
-        },
-
-        requiresPreload: {
-          type: Boolean,
-          computed: '_computeRequiresPreload(preload,loaded)'
-        },
-
-        canLoad: {
-          type: Boolean,
-          computed: '_computeCanLoad(preventLoad, src)'
-        }
+      hostAttributes: {
+        role: 'img'
       },
 
       observers: [
-        '_transformChanged(sizing, position)',
-        '_loadBehaviorChanged(canLoad, preload, loaded)',
-        '_loadStateChanged(src, preload, loaded)',
+        '_transformChanged(sizing, position)'
       ],
 
       ready: function() {
-        if (!this.hasAttribute('role')) {
-          this.setAttribute('role', 'img');
-        }
-      },
+        var img = this.$.img;
 
-      _computeImageVisibility: function() {
-        return !!this.sizing;
-      },
+        img.onload = function() {
+          var resolvedSrc = Polymer.ResolveUrl.resolveUrl(this.src, this.ownerDocument.baseURI);
+          if (img.src !== resolvedSrc) return;
 
-      _computePlaceholderVisibility: function() {
-        return !this.preload || (this.loaded && !this.fade);
+          this._setLoading(false);
+          this._setLoaded(true);
+          this._setError(false);
+        }.bind(this);
+
+        img.onerror = function() {
+          var resolvedSrc = Polymer.ResolveUrl.resolveUrl(this.src, this.ownerDocument.baseURI);
+          if (img.src !== resolvedSrc) return;
+
+          this._reset();
+
+          this._setLoading(false);
+          this._setLoaded(false);
+          this._setError(true);
+        }.bind(this);
       },
 
-      _computePlaceholderClassName: function() {
-        if (!this.preload) {
-          return '';
+      _load: function(src) {
+        if (src) {
+          this.$.img.src = src;
+        } else {
+          this.$.img.removeAttribute('src');
```

Comment:
A combination of a couple of things: setting `src` to null or undefined causes the string 'null' / 'undefined' to be written in as the src, which will call the error callback. Also, setting `src` to the empty string causes the error callback to be called in some browsers, but not all. This ended up being the only consistent way I could find to reset the state of the image.